saleor/discount/utils.py

```diff
@@ -615,11 +663,94 @@ def _clear_checkout_discount(checkout_info, save):
             )
 
 
+def _get_best_gift_reward(
+    rules: Iterable["PromotionRule"], channel: "Channel", country: str
+) -> tuple[Optional[PromotionRule], Optional[ProductVariantChannelListing]]:
+    from ..warehouse.availability import check_stock_quantity_bulk
+
+    rule_ids = [rule.id for rule in rules]
+    PromotionRuleGift = PromotionRule.gifts.through
+    rule_gifts = PromotionRuleGift.objects.filter(promotionrule_id__in=rule_ids)
+    variants = ProductVariant.objects.filter(
+        Exists(
+            rule_gifts.values("productvariant_id").filter(
+                productvariant_id=OuterRef("id")
+            )
+        )
+    )
+    variant_ids_with_insufficient_stock = set()
+    if not variants:
+        return None, None
+
+    try:
+        check_stock_quantity_bulk(
+            variants,
+            country,
+            [1] * variants.count(),
+            channel.slug,
+            None,
+        )
+    except InsufficientStock as error:
+        variant_ids_with_insufficient_stock = {
+            item.variant.pk for item in error.items if item.variant
+        }
+
+    available_variant_ids = (
+        set(variants.values_list("id", flat=True)) - variant_ids_with_insufficient_stock
+    )
+
+    if not available_variant_ids:
+        return None, None
+
+    # check if variant is available for purchase
+    available_variant_ids = _get_available_for_purchase_variant_ids(
+        available_variant_ids, channel
+    )
+    if not available_variant_ids:
+        return None, None
+
+    # check variant channel availability
+    available_variant_listings = ProductVariantChannelListing.objects.filter(
+        variant__id__in=available_variant_ids,
```

Comment:
We could use filed on `ProductVariantChannelListing` instead of ID inside related objects. 
```suggestion
        variant_id__in=available_variant_ids,
```